src/System.Net.HttpListener/tests/HttpListenerResponseTests.cs

```diff
@@ -53,13 +53,1208 @@ public void Dispose()
+            {
+                Assert.Empty(response.Headers);
+                response.Headers.Add("OldName", "OldValue");
+
+                var headers = new WebHeaderCollection
+                {
+                    { "Name1", "Value1" },
+                    { "Name2", "Value2" },
+                    { "Name3", "" }
+                };
+                response.Headers = headers;
+                Assert.Equal(headers, response.Headers);
+            }
+
+            string clientResponse = GetClientResponse();
+            Assert.Contains("\r\nName1: Value1\r\nName2: Value2\r\nName3: \r\n", clientResponse);
+        }
+
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, ".NET Core fixes a bug where HttpListenerResponse.CopyFrom(null) throws a NRE.")]
+        // The managed implementation should store ContentType in Headers, not as a separate variable.
+        [ActiveIssue(18128, TestPlatforms.AnyUnix)]
+        public async Task Headers_SetNull_ThrowsArgumentNullExceptionInNetCore()
+        {
+            using (HttpListenerResponse response = await GetResponse())
+            {
+                response.Headers.Add("OldName", "OldValue");
+
+                Assert.Throws<ArgumentNullException>("value", () => response.Headers = null);
+                Assert.Equal(1, response.Headers.Count);
+            }
+        }
+
+        [ConditionalFact(nameof(PlatformDetection) + "." + nameof(PlatformDetection.IsNotOneCoreUAP))]
+        [SkipOnTargetFramework(~TargetFrameworkMonikers.NetFramework, ".NET Core fixes a bug where HttpListenerResponse.CopyFrom(null) throws a NRE.")]
+        public async Task Headers_SetNull_ThrowsNullReferenceExceptionInNetFx()
+        {
+            using (HttpListenerResponse response = await GetResponse())
+            {
+                response.Headers.Add("OldName", "OldValue");
+
+                Assert.Throws<NullReferenceException>(() => response.Headers = null);
+                Assert.Equal(0, response.Headers.Count);
+            }
+        }
+
+        private string GetClientResponse()
+        {
+            byte[] buffer = new byte[256];
+            int bytesReceived = Client.Receive(buffer);
+            Assert.True(bytesReceived < buffer.Length, "Buffer should hold enough space to hold the response from the server.");
+
+            return Encoding.UTF8.GetString(buffer, 0, bytesReceived);
+        }
+
+        private async Task<HttpListenerResponse> GetResponse(string httpVersion = "1.1")
         {
-            Message = Client.GetAsync(Factory.ListeningUrl);
+            Client.Send(Factory.GetContent(httpVersion, "POST", null, "Give me a context, please", null, headerOnly: false));
             HttpListenerContext context = await Factory.GetListener().GetContextAsync();
             return context.Response;
         }
```

Comment:
It's the sort of testing we need to be confident we're not breaking the managed implementation!
I thought the testing guidelines are file per class? I'm happy to break up the tests though - is this because you're concerned about test execution time and want the test classes to run in parallel?